concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1161 @@
+
+Controller integration tests should be reserved for cases when we care about the eventual state of the object or need a kubenetes cluster.
+
+**See Also**
+- [kubebuilder documentation](https://book.kubebuilder.io/reference/envtest.html)
+- [envtest godoc](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest)
+
+### Unit Tests
+
+[Unit tests](https://en.wikipedia.org/wiki/Unit_testing) are used to test individual `units` of the code. In order to test the unit in an `isolated` fashion, external `dependencies` are typically `mocked`.
+Unit tests provide a very fast `feedback cycle` and help discovering bugs in a very `early stage`.
+The advantage of unit testing is that "[Code can be impossible or difficult to unit test if poorly written, thus unit testing can force developers to structure functions and objects in better ways.](https://en.wikipedia.org/wiki/Unit_testing)"
+The big disadvantage of unit tests is that they are tightly coupled to the code and a rewrite of the code may require new or modified unit tests.
+
+## Testing Guidelines 
+
+The following section describes ugly tests or problems which we discovered in our code base. The aim of the testing guidelines are to provide suggestions on how the tests can be written instead.
+They should be used whenever possible but are not set in stone. There might be cases there the proposed solution does not fit and that is ok.
+
+**Best Practices**:
+- Find a `balance` between [`KISS`](https://en.wikipedia.org/wiki/KISS_principle) (keep it simple stupid) and [`DRY`](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) (don't repeat yourself). Tests should be easily `understandable`. At the same time you should `avoid` code `duplication` because it gets un`maintaineable` very fast.
+- `Split` heavy test `setup` from actual test.
+- Test `one thing` at a time. This alignes with the KISS principle as well. Focus on the `main concern` of a test and don't test more than what you need. Comments on tests are welcome, but might also be an indication for a test which is either (1) too complex or (2) tests to many things simultaneously. 
+Concentraing on the main concern makes the individual tests shorter, easier to understand, better to maintain and expresses the intention of the test better. If multiple things are tested at the same time, there is `no guarantuee` that someone will take this secondary thing out of the test. If there are multiple tests however, you can easily see that a test case was removed and argue whether this is ok or not.
+- `Don't` test the `unforeseen`. Instead think about the code and derive useful test scenarios (best and worst case). Think about what is the `happy path` ? What is the `unhappy path` ?
+- Especially complex tests deserve some `documentation`. This helps the reader of the test to better understand it.
+
+**Goals**:
+- Avoid regression: Whenever you close a bug, you should make sure that this bug will not get reintroduced. 
+<!-- voting question:
```

Comment:
Voting:
does it make sense to:
1. use regression in the test name to indicate whether a test is a regression test ?

Reply:
2. link a PR or issue to the test so that you can understand even two years later, what the problem was at the time ?